torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +176,217 @@ def magic_impl(self, other):
-        if const_expr is not None:
-            return const_expr
-
-        expr = expr.simplify()
-        concrete_val = expr.subs(self.shape_env)
-
-        # Uncomment this to see what code triggered this guard.
-        # TODO: Save this to the guard representation so you can look
-        # at it later
-        # import traceback
-        # traceback.print_stack()
+        new_env = ShapeEnv()
+        _ = new_env.create_shapes_for_args(args)
+        return all(guard.xreplace(new_env.var_to_val) == value for guard, value, _ in self.guards)
+
+    def get_nontrivial_guards(self):
+        guards = [(self.simplify(guard), val) for guard, val, _ in self.guards]
+        guards = [guard for guard in guards if len(guard[0].free_symbols) > 0]
+        return guards
+
+    def get_shape_groups(self):
+        shape_groups = collections.defaultdict(list)
+        for k, v in self.replacements.items():
+            shape_groups[v].append(k)
+        return shape_groups
+
+    @_lru_cache
+    def _maybe_evaluate_static(self, expr: "sympy.Expr") -> "Optional[sympy.Expr]":
+        """
+        Tries to evaluate expr without introducing guards
+        """
+        # Simplifies assuming that shape vars > 1 (since we cache on 0/1 shape values)
+        symbols = list(expr.free_symbols)
+        new_shape_env = {
+            k: sympy.Symbol(f"shape_{idx}", positive=True, integer=True) + 1
+            for idx, k in enumerate(symbols)
+        }
+        new_expr = expr.xreplace(new_shape_env)
+        new_expr = sympy.expand(new_expr)
+        if len(list(new_expr.free_symbols)) == 0:
+            return new_expr
+        return None
 
-        self.guards.append((expr, concrete_val))
+    @_lru_cache
+    def replace(self, expr: "sympy.Expr") -> "sympy.Expr":
+        replacements = {s: self._find(s) for s in expr.free_symbols}
+        return sympy.expand(expr.xreplace(replacements))
+
+    @_lru_cache
+    def _update_divisible(self):
+        new_divisible = {}
+        for k in self.divisible:
+            res = self.replace(k)
+            if len(res.free_symbols) > 0:
+                new_divisible[k] = 0
+
+        self.divisible = new_divisible
+
+    @_lru_cache
+    def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+        expr = self.replace(expr)
+        if len(expr.atoms(sympy.Mod)) > 0:
+            self._update_divisible()
+            expr = expr.xreplace(self.divisible)
+            expr = sympy.expand(expr)
+        return expr
+
+    @lru_cache(256)
+    def size_hint(self, expr: "sympy.Expr"):
+        result_expr = sympy.expand(expr).xreplace(self.var_to_val)
+        assert len(result_expr.free_symbols) == 0, "Size hint has variables we don't have underlying values for"
+        return result_expr
+
+    def _find(self, a):
+        """
+        Implements a DSU to find the variable that represents a
+        TODO: Improve this to handle non-identity transitive replacements
```

Comment:
What non-identity transitive relations do we care about 🤔 . Ordering?

Reply:
I don't think we handle cases like
```
a: b + c
c: d
d: e
```

The point of this structure is so that a single xreplace is enough to completely simplify our expressions.

But this makes it not enough. I think I need to recursively call `find` on all free symbols in an expression.

Reply:
Actually just added it into this PR.

Reply:
mutable unification variables, you know you want toooooo